joinrun/src/test/scala/code/chymyst/test/Patterns01Spec.scala

```diff
@@ -101,6 +101,73 @@ class Patterns01Spec extends FlatSpec with Matchers with BeforeAndAfterEach {
     result shouldEqual ((456*456,123*123))
   }
 
+  it should "implement smokers" in {
+    val supplyLineSize = 10
+    def smoke(): Unit = Thread.sleep(math.floor(scala.util.Random.nextDouble*20.0 + 2.0).toLong)
+
+    case class ShippedInventory(tobacco: Int, paper: Int, matches: Int)
+    case class SupplyChainState(inventory: Int, shipped: ShippedInventory)
+    // this data is only to demonstrate effects of randomization on the supply chain and make content of logFile more interesting.
+    // strictly speaking all we need to keep track of is inventory.
+
+    val pusher = new M[SupplyChainState]("Pusher has delivered some unit")
+
+    val KeithInNeed = new E("Keith is in need of tobacco and matches")
+    val SlashInNeed = new E("Slash is in need of tobacco and paper")
+    val JimiInNeed = new E("Jimi is in need of matches and paper")
+
+    val tobaccoShipment = new M[SupplyChainState]("tobacco shipment")
+    val matchesShipment = new M[SupplyChainState]("matches shipment")
+    val paperShipment = new M[SupplyChainState]("paper shipment")
+
+    val pusherDone = new E("done")
+    val check = new EE("check") // blocking Unit, only blocking molecule of the example.
```

Comment:
At lines 123-124, I would just use macros `m[Unit]` and `b[Unit,Unit]` to define the molecules. But maybe you like to use the new types `E` and `EE`. I don't like them much - but right now we have to use them.